src/google/protobuf/compiler/objectivec/objectivec_helpers.cc

```diff
@@ -243,21 +243,29 @@ bool IsReservedCIdentifier(const string& input) {
   return false;
 }
 
-string SanitizeNameForObjC(const string& input,
+string SanitizeNameForObjC(const string& prefix,
+                           const string& input,
                            const string& extension,
                            string* out_suffix_added) {
   static const std::unordered_set<string> kReservedWords =
       MakeWordsMap(kReservedWordList, GOOGLE_ARRAYSIZE(kReservedWordList));
   static const std::unordered_set<string> kNSObjectMethods =
       MakeWordsMap(kNSObjectMethodsList, GOOGLE_ARRAYSIZE(kNSObjectMethodsList));
-  if (IsReservedCIdentifier(input) ||
-      (kReservedWords.count(input) > 0) ||
-      (kNSObjectMethods.count(input) > 0)) {
+  string sanitized;
+  // If the string already has a prefix, don't add it again.
+  if (prefix.length() && !HasPrefixString(input, prefix)) {
```

Comment:
So there is some risk here because a prefix of "For" with match a message of "Form" even though it clearly isn't a prefix.  To be safe, you probably want to ensure the `prefix != input` and that `input[prefix.length()]` is uppercase.

I guess the on part to debate is `prefix != input`.  As that does lead to things like _FooFoo_, but that does seems like it might be safer compared to leaving potentially generic class names like "Foo" with no way to avoid the proto class colliding with application code.

Reply:
From the examples that I ran into it seemed a lot better to clean up the duplicates when prefix == input. It made the code more readable.